### ISSUE TITLE
spans sometimes never close

### ISSUE THREAD

**Issue by tvt173** [pydantic/logfire#741]
### Description

Sometimes my spans never close. They just linger as `<ongoing?>`. This happens somewhat frequently (maybe a couple of times out of a few hundred). I see this happen both when running locally (via WSL) and on my server (linux) that I would use for production. What could be causing this?
![image](https://github.com/user-attachments/assets/bd59c472-48bb-404b-9b6e-dfb377d40fca)

Not sure if this could be part of the issue, but I do use multiprocessing in my package, and there are messages being sent from different processes concurrently.

### Python, Logfire & OS Versions, related packages (not required)

```TOML
logfire="2.11.0"
platform="Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35"
python="3.12.4 | packaged by conda-forge | (main, Jun 17 2024, 10:23:07) [GCC 12.3.0]"
[related_packages]
requests="2.32.3"
pydantic="2.8.2"
protobuf="5.29.2"
rich="13.7.1"
tomli="2.0.1"
executing="2.0.1"
opentelemetry-api="1.29.0"
opentelemetry-exporter-otlp-proto-common="1.29.0"
opentelemetry-exporter-otlp-proto-http="1.29.0"
opentelemetry-instrumentation="0.50b0"
opentelemetry-instrumentation-dbapi="0.50b0"
opentelemetry-instrumentation-jinja2="0.50b0"
opentelemetry-instrumentation-requests="0.50b0"
opentelemetry-instrumentation-sqlite3="0.50b0"
opentelemetry-instrumentation-tornado="0.50b0"
opentelemetry-instrumentation-urllib="0.50b0"
opentelemetry-instrumentation-urllib3="0.50b0"
opentelemetry-proto="1.29.0"
opentelemetry-sdk="1.29.0"
opentelemetry-semantic-conventions="0.50b0"
opentelemetry-util-http="0.50b0"
```

**Comment by alexmojaki** [comment]
The only known way for this to happen is if the process ends abruptly before the span is sent, in which case there's nothing we can do. It's possible that's not the case here and there's some bug that needs fixing but we'd need a way to reproduce and I don't see how we can start.

**Comment by tvt173** [comment]
hi @alexmojaki, thank you for your reply. please see this simplified example. i believe you should be able to replicate my issue.

```python
from functools import partial
from multiprocessing import Process, Queue
import time
from time import time, sleep
import os
import logfire

logfire.configure()

KILL_SIGNAL = "KILL-SIGNAL"
FATAL_EXCEPTION = "EXCEPTION-FATAL"

@logfire.instrument("Running worker #{worker_id}")
def worker(worker_id, task_queue, results_queue, get_jobs_func):
    """Gets the set of possible jobs to execute, then executes those (by name) it pulls from the task queue and places results on the results queue.

    Args:
        worker_id: the id for this worker
        task_queue: the task queue to pull from
        results_queue: the result queue to add results to
        get_jobs_func: the function to use to grab the local set of possible jobs to execute
        deck_attrs: attributes about the deck (used for logging only)
    """
    jobs = get_jobs_func()
    while True:
        task = task_queue.get()
        if task is None:
            # Stop signal received
            logfire.debug(f"Worker {worker_id} received stop signal.")
            results_queue.put(KILL_SIGNAL)
            logfire.debug(f"Worker {worker_id} exiting.")
            return
        else:
            task['worker_id'] = worker_id  # Add worker ID to task for logging
            job_name = task["value"]
            job = jobs[job_name]
            # print(f"Worker {worker_id} starting rule {rule_name}...")
            with logfire.span(f"Executing job {job_name}"):
                try:
                    result = job()
                    logfire.debug(f"Finished executing {job_name}")
                    results_queue.put(result)
                except Exception as e:
                    error_message = f"Error when running job {job_name}: {e}"
                    logfire.error(error_message, _exc_info=True)
                    result = (FATAL_EXCEPTION, error_message)
                    results_queue.put(result)
                    logfire.debug(f"Worker {worker_id} exiting due to error.")
                    return


def producer(tasks, task_queue, n_workers):
    """Puts tasks on the queue, and kill signals for the workers.

    Args:
        tasks: the tasks (job names) to queue
        task_queue: the queue to use
        n_workers: the number of workers dispatched
    """
    for task in tasks:
        task_queue.put(task)
    # Send stop signals to all workers
    for _ in range(n_workers):
        task_queue.put(None)  # None is the signal to stop

@logfire.instrument("Running new workload")
def main(get_jobs_func: callable, num_workers: int | None = None,) -> str:
    """Runs jobs using python's multiprocessing module via a Producer-Consumer pattern.

    Args:
        get_jobs_func: a function to get a dictionary of jobs (by unique name) to be run
        num_workers: the number of workers to employ. Defaults to None, which will use the number of system cores available.
        deck_attrs: attributes about the deck, used for logging

    Returns:
        The result database, as a string.
    """
    if num_workers is None:
        num_workers = len(os.sched_getaffinity(0))
        print(f"Number of workers not set. Defaulting to {num_workers}")
    start = time()

    workers = {}

    # queue for tasks to be completed by workers
    task_queue = Queue()
    # queue for results
    results_queue = Queue()
    # Create queues and start worker processes
    for worker_id in range(num_workers):
        worker_proc = Process(target=worker, args=(worker_id, task_queue, results_queue, get_jobs_func))
        worker_proc.start()
        workers[worker_id] = worker_proc

    # create tasks
    jobs = get_jobs_func()
    tasks = [{'task_id': i_task, 'value': job_name} for i_task, job_name in enumerate(jobs.keys())]


    # Start the producer
    producer(tasks, task_queue, num_workers)

    # Consume results while workers are running
    finished_workers = 0
    results = []
    while finished_workers < num_workers:
        result = results_queue.get()
        if result == KILL_SIGNAL:
            finished_workers += 1
        elif isinstance(result, tuple) and len(result) == 2 and result[0] == FATAL_EXCEPTION:
            error_message = result[1]
            for wid, wproc in workers.items():
                if wproc.is_alive():
                    wproc.terminate()
                    wproc.join()

            logfire.error("Aborting workload due to worker error.")
            raise RuntimeError(error_message)
        else:
            results.append(result)

    # Wait for all workers to finish
    for worker_id, worker_proc in workers.items():
        worker_proc.join()

    merged_result = '\n'.join(results)
    elapsed = time() - start
    print(f"All workers have finished in {elapsed} seconds.")
    return merged_result

def sleep_and_say_hi(seconds: float):
    sleep(seconds)
    return f"I feel rested now... I slept for {seconds} s"
    
def get_jobs():
    jobs = {i: partial(sleep_and_say_hi, seconds=i/50.0) for i in range(50)}
    return jobs

if __name__ == '__main__':
    # run the workload
    main(get_jobs_func=get_jobs, num_workers=30)
```

**Comment by tvt173** [comment]
really? it pretty reliably looks like this for me:
![image](https://github.com/user-attachments/assets/687e459e-3ab3-4609-9ca8-07f54339af08)

note: the whole workload actually finished before i even opened the live view, in only 1.5 seconds

**Comment by tvt173** [comment]
and if i come back to it later, it looks like this:
![image](https://github.com/user-attachments/assets/38848866-628a-4159-9a02-3787c56c284c)

**Comment by alexmojaki** [comment]
This is what it looks like for me:

<img width="1133" alt="Screenshot 2025-01-06 at 13 17 38" src="https://github.com/user-attachments/assets/6b639e0f-ac2d-4c58-a1c5-d4ae1247ccf1" />

It doesn't look like we're actually running the same code, because "Running new workload" isn't a parent of the worker spans for me, and I don't see any reason why it would be because there's no context propagation in your code.

**Comment by tvt173** [comment]
hmm... i _am_ running that same code. it must be something in my environment then? could one of the "related packages" i have installed (as listed in the issue description) be causing this? i will have to take a look tomorrow...

btw, the desired behavior would be that the "Running new workload" span is the parent of the workers. i haven't done any explicit context propagation though (not sure how to), so i'm not sure why i'm seeing this on my end...

**Comment by alexmojaki** [comment]
@Kludex what do you get on your machine?

**Comment by Kludex** [comment]
<img width="1503" alt="Screenshot 2025-01-06 at 14 58 07" src="https://github.com/user-attachments/assets/f2c94e0f-586f-4da3-8e43-bb0ccb1dbf5d" />

**Comment by tvt173** [comment]
odd... i tried to run in a simplified environment, created with simply `conda create -n logfire-debug python=3.12 logfire`

which gives me this environment.yml
```
name: logfire-debug
channels:
  - conda-forge
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=2_gnu
  - backoff=2.2.1=pyhd8ed1ab_1
  - brotli-python=1.1.0=py312h2ec8cdc_2
  - bzip2=1.0.8=h4bc722e_7
  - ca-certificates=2024.12.14=hbcca054_0
  - certifi=2024.12.14=pyhd8ed1ab_0
  - cffi=1.17.1=py312h06ac9bb_0
  - charset-normalizer=3.4.1=pyhd8ed1ab_0
  - deprecated=1.2.15=pyhd8ed1ab_1
  - executing=2.1.0=pyhd8ed1ab_1
  - googleapis-common-protos=1.66.0=pyhff2d567_0
  - h2=4.1.0=pyhd8ed1ab_1
  - hpack=4.0.0=pyhd8ed1ab_1
  - hyperframe=6.0.1=pyhd8ed1ab_1
  - idna=3.10=pyhd8ed1ab_1
  - importlib-metadata=8.5.0=pyha770c72_1
  - ld_impl_linux-64=2.43=h712a8e2_2
  - libexpat=2.6.4=h5888daf_0
  - libffi=3.4.2=h7f98852_5
  - libgcc=14.2.0=h77fa898_1
  - libgcc-ng=14.2.0=h69a702a_1
  - libgomp=14.2.0=h77fa898_1
  - liblzma=5.6.3=hb9d3cd8_1
  - libnsl=2.0.1=hd590300_0
  - libsqlite=3.47.2=hee588c1_0
  - libstdcxx=14.2.0=hc0a3c3a_1
  - libstdcxx-ng=14.2.0=h4852527_1
  - libuuid=2.38.1=h0b41bf4_0
  - libxcrypt=4.4.36=hd590300_1
  - libzlib=1.3.1=hb9d3cd8_2
  - logfire=2.11.1=pyhd8ed1ab_0
  - markdown-it-py=3.0.0=pyhd8ed1ab_1
  - mdurl=0.1.2=pyhd8ed1ab_1
  - ncurses=6.5=he02047a_1
  - openssl=3.4.0=h7b32b05_1
  - opentelemetry-api=1.29.0=pyhd8ed1ab_1
  - opentelemetry-exporter-otlp-proto-common=1.29.0=pyhd8ed1ab_0
  - opentelemetry-exporter-otlp-proto-http=1.29.0=pyhd8ed1ab_1
  - opentelemetry-instrumentation=0.50b0=pyhd8ed1ab_1
  - opentelemetry-proto=1.29.0=pyhd8ed1ab_0
  - opentelemetry-sdk=1.29.0=pyhd8ed1ab_0
  - opentelemetry-semantic-conventions=0.50b0=pyh3cfb1c2_0
  - packaging=24.2=pyhd8ed1ab_2
  - pip=24.3.1=pyh8b19718_2
  - protobuf=5.28.3=py312h2ec8cdc_0
  - pycparser=2.22=pyh29332c3_1
  - pygments=2.18.0=pyhd8ed1ab_1
  - pysocks=1.7.1=pyha55dd90_7
  - python=3.12.8=h9e4cc4f_1_cpython
  - python_abi=3.12=5_cp312
  - readline=8.2=h8228510_1
  - requests=2.32.3=pyhd8ed1ab_1
  - rich=13.9.4=pyhd8ed1ab_1
  - setuptools=75.6.0=pyhff2d567_1
  - tk=8.6.13=noxft_h4845f30_101
  - tomli=2.2.1=pyhd8ed1ab_1
  - typing-extensions=4.12.2=hd8ed1ab_1
  - typing_extensions=4.12.2=pyha770c72_1
  - tzdata=2024b=hc8b5060_0
  - urllib3=2.3.0=pyhd8ed1ab_0
  - wheel=0.45.1=pyhd8ed1ab_1
  - wrapt=1.17.0=py312h66e93f0_0
  - zipp=3.21.0=pyhd8ed1ab_1
  - zstandard=0.23.0=py312hef9b889_1
  - zstd=1.5.6=ha6fb4c9_0
```

i was thinking this might allow me to reproduce your results, but i still get this
![image](https://github.com/user-attachments/assets/137ca488-54a4-48e9-950e-fe18ae354742)

i also confirmed that i am running exactly the same script i sent in the comment above

any ideas why this might be or what else i can try?

**Comment by alexmojaki** [comment]
Can you reduce the code to a minimal example that reproduces the problem on your side? Does this do it?

```python
from multiprocessing import Process
from time import sleep

import logfire

logfire.configure()


@logfire.instrument('Running worker #{worker_id}')
def worker(worker_id):
    sleep(worker_id)


@logfire.instrument('Running new workload')
def main():
    num_workers = 3
    workers = {}
    for worker_id in range(num_workers):
        worker_proc = Process(target=worker, args=(worker_id,))
        worker_proc.start()
        workers[worker_id] = worker_proc

    for worker_id, worker_proc in workers.items():
        worker_proc.join()


if __name__ == '__main__':
    main()
```

**Comment by tvt173** [comment]
hi @alexmojaki , thanks for the suggestion. indeed i still get the problem with your simplified example:
![image](https://github.com/user-attachments/assets/4e55062c-9f20-4f68-a562-eeb900bbd4cf)

to be sure, i also tried reducing the example even further (running a simple loop with no multiprocessing). that indeed works as expected. so it definitely seems to come from how the job is handled on the child processes.

if i were to properly propagate the context, how should this example look? i wonder if that might fix it?

**Comment by tvt173** [comment]
interesting... i lose `#0` every time it seems...
![image](https://github.com/user-attachments/assets/0ba079a8-53f6-4a28-bd74-322066d44b4a)

**Comment by tvt173** [comment]
added a few more logging messages. i see them for worker 1 and 2, but nothing for worker 0.
![image](https://github.com/user-attachments/assets/375d3ff6-21c2-46c5-8775-c59870966c2a)

in my local terminal however, i see all of them:
![image](https://github.com/user-attachments/assets/6371c92e-22c5-4ff1-b38b-80ad0968bf7c)

**Comment by alexmojaki** [comment]
```python
from multiprocessing import Process
from time import sleep

import logfire

logfire.configure()


def worker(worker_id):
    with logfire.span(f'Running worker #{worker_id}'):
        sleep(worker_id)
    logfire.force_flush()


@logfire.instrument('Running new workload')
def main():
    num_workers = 3
    workers = {}
    for worker_id in range(num_workers):
        worker_proc = Process(target=worker, args=(worker_id,))
        worker_proc.start()
        workers[worker_id] = worker_proc

    for worker_id, worker_proc in workers.items():
        worker_proc.join()


if __name__ == '__main__':
    main()
```

(I added `logfire.force_flush()` to the end of `worker`)

**Comment by tvt173** [comment]
it works!

first workload is the script with `force_flush()` and the second is with that line commented
![image](https://github.com/user-attachments/assets/78681d10-1671-4e2a-9594-1d0ab9eebca4)

so that means that the messages were getting buffered and not sent before joining the worker processes? however only the message sent to the server i presume, since the local logging worked as expected...

i still don't understand why the spans would fall into a tree structure when run from my machine but not yours though... but anyways, happy to have a workaround. thanks!

**Comment by alexmojaki** [comment]
https://github.com/pydantic/logfire/issues/779

**Comment by alexmojaki** [comment]
> btw, the desired behavior would be that the "Running new workload" span is the parent of the workers. i haven't done any explicit context propagation though (not sure how to), so i'm not sure why i'm seeing this on my end...

Here's how to do this without the fork method: https://logfire.pydantic.dev/docs/how-to-guides/distributed-tracing/#manual-context-propagation

We could maybe make `Process.start` do this by default, but in practice I don't think you'd often put a span around the entire process lifecycle, but rather individual operations. So you'd want to put those context dicts in the queue to link consumer and producer spans for each task.

**Comment by tvt173** [comment]
btw, after a few more days of wrestling with this, i highly recommend using `set_start_method('spawn')`, as you previously recommended. with forking, some jobs would hang indefinitely when trying to join the worker processes, even after your fixes in 3.1.0. with spawn, the runtime increases a bit, but it works consistently
